geoopt/optim/rwolfe.py

```diff
@@ -0,0 +1,249 @@
+        self.fallback_stepsize = self.param_groups[0]["fallback_stepsize"]
+        self.amax = self.param_groups[0]["amax"]
+        self.amin = self.param_groups[0]["amin"]
+        self.old_phi = None
+        self.step_size_history = []
+        self.closure = None
+        self._step_size_dic = dict()
+
+    def phi_(self, step_size):
+        """Compute the line search objective, and store its derivatives in the state."""
+
+        if step_size in self._step_size_dic:
+            return self._step_size_dic[step_size]
+
+        param_copy = [param.clone() for param in self._params]
+
+        for point in self._params:
+            state = self.state[point]
+            if "grad" not in state:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            grad = state["grad"]
+
+            # compute retract and transport in gradient direction
+            new_point, grad_retr = manifold.retr_transp(point, -step_size * grad, grad)
+            with torch.no_grad():
+                point.copy_(new_point)
+
+            state["grad_retr"] = grad_retr
+
+        # recompute loss at new point
+        phi = self.closure()
+
+        # Store new gradients in state
+        for point in self._params:
+            grad = point.grad
+            if grad is None:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+
+            state["new_grad"] = grad
+
+        # roll back parameters to before step
+        with torch.no_grad():
+            for point, old_point in zip(self._params, param_copy):
+                point.copy_(old_point)
+
+        self._step_size_dic[step_size] = phi
+        return phi
+
+    def derphi_(self, step_size):
+        """Compute derivative of phi.
+
+        The derivative of phi is given by computing inner
+        product between all tensor gradients at target point and those at source point.
+        The source gradients are transported to the target point, and both gradients are
+        projected.
+        """
+
+        # Call phi_ to compute gradients; Does nothing if phi_ was
+        # already called with this stepsize during this step
+        self.phi_(step_size)
+
+        derphi = 0
+        for point in self._params:
+            state = self.state[point]
+            if "grad" not in state:
+                continue
+
+            # For some reason using the metric for inner product gives wrong results,
+            # therefore us euclidean product instead.
```

Comment:
what was wrong there?

Reply:
I now realized what went wrong probably. If using `manifold.inner(point, -state["new_grad"], state["grad_retr"])` at this point in the code, then the point is wrong. We need to compute this quantity before we roll back point at the end of the function `_phi`, and doing this seems to work well. I will modify the code.